lang/src/org/partiql/lang/eval/visitors/AggregateSupportVisitorTransform.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ * A copy of the License is located at:
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+ *  language governing permissions and limitations under the License.
+ */
+
+package org.partiql.lang.eval.visitors
+
+import com.amazon.ionelement.api.MetaContainer
+import com.amazon.ionelement.api.metaContainerOf
+import org.partiql.lang.ast.AggregateCallSiteListMeta
+import org.partiql.lang.ast.AggregateRegisterIdMeta
+import org.partiql.lang.ast.SelectProjectionValue
+import org.partiql.lang.domains.PartiqlAst
+
+/**
+ * Allocates registerIds to all aggregate call-sites, storing the allocated registerId in an instance of
+ * [AggregateRegisterIdMeta].
+ *
+ * Also collects a list of all aggregate functions used in the `SELECT` list of the current query (excluding subqueries)
+ * and stores them in an instance of [AggregateCallSiteListMeta] on the [PartiqlAst.Expr.Select] instance.
+ */
+
+class AggregateSupportVisitorTransform : PartiqlAst.VisitorTransform() {
+    private val aggregateCallSites = ArrayList<PartiqlAst.Expr.CallAgg>()
+
+    inner class RegisterIdAdderSubVisitorTransform : PartiqlAst.VisitorTransform() {
+        /**
+         * Nests another [AggregateSupportVisitorTransform] within this transform
+         * in order to avoid exposing [aggregateCallSites] to subqueries.
+         */
+        override fun transformExprSelect(node: PartiqlAst.Expr.Select): PartiqlAst.Expr =
+            AggregateSupportVisitorTransform().transformExpr(node)
+
+        override fun transformExprCallAgg(node: PartiqlAst.Expr.CallAgg): PartiqlAst.Expr {
+            aggregateCallSites.add(node)
+
+            return PartiqlAst.build {
+                callAgg_(
+                    setq = node.setq,
+                    funcName = node.funcName,
+                    arg = transformExpr(node.arg),
+                    metas = transformMetas(node.metas) + metaContainerOf(AggregateRegisterIdMeta.TAG to AggregateRegisterIdMeta(aggregateCallSites.size - 1))
```

Comment:
To fix this [test](https://github.com/partiql/partiql-lang-kotlin/blob/774863f73d06f7065166801c62bff4bff46c56e0/lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt#L62), had to change to use `node.metas` rather than `node.arg.metas`. It makes more sense that the `CallAgg`'s source location meta use the source location of where the aggregation function starts rather than where the argument to the function starts.